# (PART) WHY DOCKER {-}

# Why Using Docker

Docker has revolutionized the way omics data analysis workflows are developed, shared, and deployed. Omics data, which includes genomics, proteomics, metabolomics, and other high-dimensional biological datasets, often requires complex software environments and dependencies. Docker provides a solution by encapsulating these environments into portable and reproducible containers.

In this guide, we'll explore how Docker can enhance omics data analysis workflows:

## Creating Reproducible Environments
Docker allows researchers to package their entire analysis pipeline, including software tools, libraries, and dependencies, into a single container. This ensures reproducibility across different computing environments, making it easier to share and reproduce analyses.

## Simplified Installation
Setting up bioinformatics software tools and libraries can be challenging due to compatibility issues and dependency conflicts. With Docker, users can simply pull pre-built Docker images containing all the necessary software, eliminating the need for manual installations.

## Scalability and Efficiency
Docker containers are lightweight and can be spun up or down quickly, making them ideal for parallelizing omics analyses across multiple computing nodes or cloud instances. This scalability improves efficiency and reduces analysis time for large datasets.

## Version Control and Collaboration
Docker images can be version-controlled using tools like Docker Hub or Git, allowing researchers to track changes and collaborate on analysis pipelines. This facilitates reproducible research and enables seamless sharing of analysis workflows with the scientific community.

## Integration with Workflow Managers
Docker integrates seamlessly with workflow management systems like Nextflow, Snakemake, or Common Workflow Language (CWL). Researchers can define complex analysis pipelines using these workflow managers and execute them within Docker containers, ensuring consistency and reproducibility.

By leveraging Docker for omics data analysis, researchers can streamline their workflows, improve reproducibility, and accelerate scientific discoveries in the field of bioinformatics. In the following sections, we'll dive deeper into the practical aspects of using Docker for various omics data analysis tasks, from preprocessing raw data to downstream statistical analysis and visualization. Let's embark on a journey to harness the power of Docker for advancing omics research!

# (PART) GETTING STARTED {-}

# Creating a Docker Account

## Go to Docker Hub
- **Step 1**: Open your web browser and navigate to the Docker Hub website: [hub.docker.com](https://hub.docker.com/).

## Sign Up
- **Step 2**: On the Docker Hub homepage, you'll find a "Sign Up" button at the top right corner. Click on it.

## Fill in Details
- **Step 3**: You'll be prompted to enter your details for creating an account. This typically includes your username, email address, and password. Fill in the required fields.

## Agree to Terms
- **Step 4**: After filling in your details, you may need to agree to the Docker Terms of Service and Privacy Policy. Read through them and then check the box to agree.

## Complete Sign Up
- **Step 5**: Once you've filled in your details and agreed to the terms, click on the "Sign Up" button to create your Docker account.

## Verify Email (if required)
- **Step 6**: Depending on Docker's current policies, you may need to verify your email address by clicking on a link sent to the email you provided during sign-up. Check your email inbox and follow the instructions provided to verify your email address.

## Access Your Account
- **Step 7**: After completing the sign-up process and email verification (if required), you can now log in to your Docker account using the username and password you provided during sign-up.

## Explore Docker Hub
- **Step 8**: Once logged in, you can explore Docker Hub to search for existing Docker images, pull images to your local machine, push your own images to Docker Hub, and manage your repositories.

By following these steps, you'll have successfully created a Docker account and gained access to Docker Hub, where you can leverage a vast collection of container images and collaborate with the Docker community.


# Using a Docker Image from Docker Hub

1. **Search for the Image**: Visit the [Docker Hub](https://hub.docker.com/) website and search for the Docker image you want to use.

2. **Pull the Image**: Once you've found the desired Docker image, pull it to your local machine using the following command:


# (PART) USE CASE {-}

# Use Case: Junction Analysis Using Portcullis Docker Image
> Docker reference: https://www.digitalocean.com/community/tutorials/how-to-remove-docker-images-containers-and-volumes

## Step 1: Pull or build image from a `Dockerfile`
```bash
docker pull maplesond/portcullis:latest
docker pull 17602532556/annotation:latest
```

## Step 2: Run and mount a volume
- Run the pulled image by mounting it to e.g \data. 
- Then try to ls to see if you can see the files in the mounted directory

```bash
docker run -t -i -v $(pwd):/data maplesond/portcullis:latest
```
> Output is a Linux CLI: `/data # `

## Step 3: Run all Portcullis Steps at Once
- Using the `full` subtool

```bash
for i in data/*.bam; do portcullis full -t 4 -o ${i}.out data/C_auris_B8441_current_chromosomes.fasta $i; done
```

> Make sure we have a sufficent number of junctions to work with.  Minimum 1000

> Not enough junctions to create training set.

# Step 4: Junctions Annotation 
- Using Redtools

## Prepare a bed file
- Use `regtools junctions extract [options] indexed_alignments.bam`
```bash
regtools junctions annotate -S -o redtools_out/annotated.junctions.tab portcullis_out/3-filt/portcullis_filtered.pass.junctions.bed data/C_auris_B8441_current_chromosomes.fasta data/C_auris_B8441_current_features.gtf

```

```bash
for i in data/*.bam; do portcullis full -t 4 -o ${i}.out data/C_auris_B8441_current_chromosomes.fasta $i; done
```

```bash
for i in data/*.out; do regtools junctions annotate -S -o ${i}.annotated.junctions.tab $i/3-filt/portcullis_filtered.pass.junctions.bed data/C_auris_B8441_current_chromosomes.fasta data/C_auris_B8441_current_features.gtf; done
```


